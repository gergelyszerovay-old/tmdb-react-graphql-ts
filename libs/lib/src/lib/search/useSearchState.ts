import React from "react";
import {movieListTermVar, similarMoviesListVar} from "../state-gql/cache";

export interface SearchState {
  onQueryMovieList: () => void,
  inputTerm: string,
  setInputTerm: (string) => void
}

const useSearchState = (): SearchState => {
  const [inputTerm, setInputTerm] = React.useState("");

  return {
    onQueryMovieList: () => {
      movieListTermVar(inputTerm);
      similarMoviesListVar(false);
    },
    inputTerm, setInputTerm
  }
};

export default useSearchState;
