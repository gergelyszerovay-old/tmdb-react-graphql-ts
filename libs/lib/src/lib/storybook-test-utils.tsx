import React from 'react';
import {StylesProvider} from '@material-ui/styles';

// https://github.com/mui-org/material-ui/issues/9492
export const WithStableMuiClassnames = (c) => {
  const generateClassName = () => {
    let counter = 0
    return (rule, styleSheet) => `${styleSheet.options.classNamePrefix}-${rule.key}-${counter++}`
  }

  return (
    <StylesProvider generateClassName={generateClassName()}>
      {c}
    </StylesProvider>
  )
}
