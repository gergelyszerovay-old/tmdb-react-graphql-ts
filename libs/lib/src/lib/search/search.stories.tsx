import React from 'react';
import {Search} from './search';
import {action} from "@storybook/addon-actions";
import useSearchState, {SearchState} from "./useSearchState";

export default {
  component: Search,
  title: 'Search',
};

export const primary = () => {
  const useMockState = (): SearchState => {
    return {
      ...useSearchState(),
      onQueryMovieList: action('onQueryMovieList')
    };
  };

  return <Search hook={useMockState}/>;
};
