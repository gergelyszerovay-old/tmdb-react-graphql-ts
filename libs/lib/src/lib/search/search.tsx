import React, {FC} from 'react';

import './search.scss';
import {AppBar, createStyles, fade, InputBase, Theme, Toolbar} from "@material-ui/core";
import SearchIcon from '@material-ui/icons/Search';
import Button from "@material-ui/core/Button";
import {makeStyles} from "@material-ui/core/styles";
import useSearchState, {SearchState} from "./useSearchState";

const useStyles = makeStyles((theme: Theme) =>
  // styles based on https://material-ui.com/components/app-bar/
  createStyles({
    search: {
      position: 'relative',
      borderRadius: theme.shape.borderRadius,
      backgroundColor: fade(theme.palette.common.white, 0.15),
      '&:hover': {
        backgroundColor: fade(theme.palette.common.white, 0.25),
      },
      marginRight: theme.spacing(2),
      marginLeft: 0,
      width: '100%',
      [theme.breakpoints.up('sm')]: {
        marginLeft: theme.spacing(3),
        width: 'auto',
      },
    },
    searchIcon: {
      padding: theme.spacing(0, 2),
      height: '100%',
      position: 'absolute',
      pointerEvents: 'none',
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    },
    inputRoot: {
      color: 'inherit',
    },
    inputInput: {
      padding: theme.spacing(1, 1, 1, 0),
      // vertical padding + font size from searchIcon
      paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
      transition: theme.transitions.create('width'),
      width: '100%',
      [theme.breakpoints.up('md')]: {
        width: '20ch',
      },
    },
  }),
);

/* eslint-disable-next-line */
export interface SearchProps {
  hook?: () => SearchState
}

export const Search: FC<SearchProps> = ({hook = useSearchState}: SearchProps) => {
  const classes = useStyles();

  const {inputTerm, setInputTerm, onQueryMovieList} = hook();

  return (
    <div>
      <AppBar position="static">
        <Toolbar>
          <div className={classes.search}>
            <div className={classes.searchIcon}>
              <SearchIcon/>
            </div>
            <InputBase
              placeholder="Searchâ€¦"
              classes={{
                root: classes.inputRoot,
                input: classes.inputInput,
              }}
              inputProps={{'aria-label': 'search', 'data-testid': 'search-input'}}
              value={inputTerm}
              onChange={e => setInputTerm(e.target.value)}
              onKeyPress={e => {
                if (e.charCode === 13) {
                  onQueryMovieList();
                }
              }
              }
            />
          </div>
          <Button
            variant="contained"
            color="secondary"
            onClick={() => onQueryMovieList()}
            data-testid="search-button"
          >
            Search
          </Button>
        </Toolbar>
      </AppBar>
    </div>
  );
};

export default Search;
