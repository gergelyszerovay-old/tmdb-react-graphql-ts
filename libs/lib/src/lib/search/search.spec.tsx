import React from 'react';
import {fireEvent, getByTestId, render} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Search from './search';
import * as Stories from "./search.stories";
import useSearchState, {SearchState} from "./useSearchState";
import {WithStableMuiClassnames} from "../storybook-test-utils";

describe('Search', () => {

  describe('Search field submit tests', () => {
    let container = null;
    let onQueryMovieListMock = null;
    let state: SearchState = null;

    beforeEach(() => {
      onQueryMovieListMock = jest.fn();
      const useMockState = (): SearchState => {
        state = useSearchState();
        return {
          ...state,
          onQueryMovieList: onQueryMovieListMock
        };
      };

      container = render(<Search hook={useMockState}/>).container;
    })

    it('Submit on enter should work', () => {
      const searchInputEl = getByTestId(container, 'search-input') as HTMLInputElement;
      userEvent.type(searchInputEl, 'Test123');
      fireEvent.keyPress(searchInputEl, {key: 'Enter', code: 'Enter', charCode: 13});
      expect(onQueryMovieListMock).toHaveBeenCalled();
      expect(state.inputTerm).toBe('Test123');
    });

    it('Click on search button should work', () => {
      const searchInputEl = getByTestId(container, 'search-input') as HTMLInputElement;
      const searchButtonEl = getByTestId(container, 'search-button') as HTMLInputElement;
      userEvent.type(searchInputEl, 'Test123');
      fireEvent.click(searchButtonEl);
      expect(onQueryMovieListMock).toHaveBeenCalled();
      expect(state.inputTerm).toBe('Test123');
    });
  });

  describe('Storybook stories snapshot tests', () => {
    for (const [key, value] of Object.entries(Stories)) {
      if (value instanceof Function) {
        it(`Story ${key}`, () => {
          const {container} = render(WithStableMuiClassnames(Stories[key]()));
          expect(container.firstChild).toMatchSnapshot()
        });
      }
    }
  });

});
