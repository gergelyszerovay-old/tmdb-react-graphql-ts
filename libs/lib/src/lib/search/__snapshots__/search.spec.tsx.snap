// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Storybook stories snapshot tests Story primary 1`] = `
<div>
  <header
    class="MuiPaper-root-15 MuiAppBar-root-4 MuiAppBar-positionStatic-8 MuiAppBar-colorPrimary-11 MuiPaper-elevation4-22"
  >
    <div
      class="MuiToolbar-root-43 MuiToolbar-regular-45 MuiToolbar-gutters-44"
    >
      <div
        class="makeStyles-search-0"
      >
        <div
          class="makeStyles-searchIcon-1"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-47"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
            />
          </svg>
        </div>
        <div
          class="MuiInputBase-root-56 makeStyles-inputRoot-2"
        >
          <input
            aria-label="search"
            class="MuiInputBase-input-67 makeStyles-inputInput-3"
            data-testid="search-input"
            placeholder="Searchâ€¦"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root-103 MuiButton-root-74 MuiButton-contained-82 MuiButton-containedSecondary-84"
        data-testid="search-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-75"
        >
          Search
        </span>
        <span
          class="MuiTouchRipple-root-106"
        />
      </button>
    </div>
  </header>
</div>
`;
